src/mappings/storage/utils.ts

```diff
@@ -206,3 +214,134 @@ export async function deleteDataObjectsByIds(
   subtitlesRepository.remove(...currentSubtitles.flat())
   await deleteDataObjects(overlay, block, indexInBlock, extrinsicHash, objects)
 }
+
+export async function processSetNodeOperationalStatusMessage(
+  overlay: EntityManagerOverlay,
+  block: Block,
+  indexInBlock: number,
+  extrinsicHash: string | undefined,
+  workingGroup: 'storageWorkingGroup' | 'distributionWorkingGroup',
+  actor: Worker | undefined, // worker who sent the message, undefined if it's a lead
+  meta: ISetNodeOperationalStatus
+): Promise<void> {
+  let workerId: bigint
+  const bucketId = meta.bucketId || ''
+
+  // Get the bucket's worker id
+  if (actor) {
+    workerId = actor.runtimeId
+  } else {
+    const maybeWorker = await overlay
+      .getRepository(Worker)
+      .getById(`${workingGroup}-${meta.workerId}`)
+
+    if (!maybeWorker) {
+      return invalidMetadata(
+        SetNodeOperationalStatus,
+        `The worker ${meta.workerId} does not exist in the ${workingGroup} working group`
+      )
+    }
+    workerId = maybeWorker.runtimeId
+  }
+
+  // Update the operational status of Storage node
+  if (workingGroup === 'storageWorkingGroup') {
+    const storageBucket = await overlay.getRepository(StorageBucket).getById(bucketId)
+    if (!storageBucket) {
+      return invalidMetadata(
+        SetNodeOperationalStatus,
+        `The storage bucket ${bucketId} does not exist`
+      )
+    } else if (storageBucket.operatorStatus.isTypeOf !== 'StorageBucketOperatorStatusActive') {
+      return invalidMetadata(
+        SetNodeOperationalStatus,
+        `The storage bucket ${bucketId} is not active`
+      )
+      // If the actor is a worker, check if the worker is the operator of the storage bucket
+    } else if (actor && storageBucket.operatorStatus.workerId !== actor.runtimeId) {
+      return invalidMetadata(
+        SetNodeOperationalStatus,
+        `The worker ${workerId} is not the operator of the storage bucket ${bucketId}`
+      )
+    }
+
+    // create metadata entity if it does not exist already
+    const metadataEntity =
+      (await overlay.getRepository(StorageBucketOperatorMetadata).getById(bucketId)) ||
+      overlay
+        .getRepository(StorageBucketOperatorMetadata)
+        .new({ id: bucketId, storageBucketId: bucketId })
+
+    if (isSet(meta.operationalStatus)) {
+      metadataEntity.nodeOperationalStatus = processNodeOperationalStatusMetadata(
+        actor ? 'worker' : 'lead',
+        metadataEntity.nodeOperationalStatus,
+        meta.operationalStatus
+      )
+    }
+
+    // event processing
+
+    const operationalStatusSetEvent = new StorageNodeOperationalStatusSetEvent({
+      ...genericEventFields(overlay, block, indexInBlock, extrinsicHash),
+      storageBucket: storageBucket.id,
+      operationalStatus: metadataEntity.nodeOperationalStatus || undefined,
```

Comment:
Here we are emitting `metadataEntity.nodeOperationalStatus`, will this be already updated value after call to `processNodeOperationalStatusMetadata`?

Reply:
It will be set on the line.

```ts
metadataEntity.nodeOperationalStatus = processNodeOperationalStatusMetadata(
```

But I just noticed that this `StorageNodeOperationalStatusSetEvent` evet is being emitted regardless of the outcome of the  `processNodeOperationalStatusMetadata`. 

So, I have changed the implementation only to emit the event when the status changes.